model/task/task.go

```diff
@@ -658,9 +661,68 @@ func (t *Task) MarkAsDispatched(hostId, distroId, agentRevision string, dispatch
 			return t.DisplayTask.MarkAsDispatched("", "", "", dispatchTime)
 		}
 	}
+
 	return nil
 }
 
+func (t *Task) CalculateUnblockedTime() error {
+	//get all immediate dependencies
+	directDependencies := t.DependsOn
+
+	//if dependencies exist, get the max of their finish times
+	if len(directDependencies) > 0 {
+		grip.Debug(message.Fields{"message": "unblocked time debug", "task_id": t.Id, "event": "Entered for loop"})
+		latestTime := t.ScheduledTime
+		for _, dep := range directDependencies {
+			depID := dep.TaskId
+
+			depTask, err := FindOne(ById(depID).WithFields(FinishTimeKey, ExecutionKey))
+
+			if err != nil {
+				grip.Debug(message.Fields{"message": "unblocked time debug", "task_id": t.Id, "event": err})
+				return err
+			}
+
+			if depTask.Execution > 0 {
+				// we can't calculate unblocked time without looking at
+				// oldtasks collection, so we won't calculate at all
+				return nil
+			}
+			if depTask.FinishTime.After(latestTime) {
+				latestTime = depTask.FinishTime
+			}
+			if depTask.FinishTime.After(t.StartTime) {
+				return errors.Errorf(
+					"unexpected task finish time %s after dependent start time %s",
+					depTask.FinishTime.String(), t.StartTime.String())
+			}
+
+		}
+		// record in the in-memory task
+		t.UnblockedTime = latestTime
+		// update the db
+		err := UpdateOne(
+			bson.M{
+				IdKey: t.Id,
+			},
+			bson.M{
+				"$set": bson.M{
+					LastHeartbeatKey: latestTime,
+					UnblockedTimeKey: latestTime,
+				},
+			},
+		)
+		if err != nil {
+			grip.Debug(message.Fields{"message": "unblocked time debug", "task_id": t.Id, "event": err})
+			return err
+		}
+
+		return nil
+	} else {
```

Comment:
Since no work happens if `len(t.DependsOn) == 0`, it's better style to just return immediately in that case, and then the rest of the function doesn't have to be in this if block